### PR TITLE
fix: handle partial reads in UnpadReader for non-power-of-2 pieces

### PR DESCRIPTION
**Pull request by rvagg**
Ref: https://filecoinproject.slack.com/archives/CP50PPW2X/p1756905471593619
Ref: https://github.com/filecoin-project/boost/pull/2020

- Keep PR #12884's buffer overflow fix (clamp to buffer size)
- Add graceful handling of io.ErrUnexpectedEOF for partial pieces
- Fixes boost EOF errors when reading 7MB deals in 8MB sectors
- Add test for partial piece reading edge case

### CODE REVIEW

File: storage/sealer/fr32/readers.go [hunk 1]
```diff
@@ -101,11 +101,37 @@ func (r *unpadReader) readInner(out []byte) (int, error) {
...
+		// doesn't have as much data as expected (e.g., non-power-of-2 pieces).
+		// Process what we got.
+		if n > 0 {
+			// Round down to complete 128-byte chunks
+			completeChunks := n / 128
+			if completeChunks > 0 {
+				validBytes := completeChunks * 128
+				Unpad(r.work[:validBytes], out[:completeChunks*127])
+				// Adjust left to reflect that we couldn't read everything
+				r.left = 0
```
[the remark this answers is not in the document]
Comment (rvagg) [reply storage/sealer/fr32/readers.go:115]: we got an `io.ErrUnexpectedEOF` here, so there's no more reads

File: storage/sealer/fr32/readers_test.go [hunk 2]
```diff
@@ -196,3 +196,50 @@ func TestUnpadReaderLargeReads(t *testing.T) {
...
 }
+
+// TestUnpadReaderPartialPiece tests the edge case where the underlying reader
+// has less data than UnpadReader expects. This happens when dealing with pieces
+// that are not exact power-of-2 sizes.
+// This test captures the fix for the EOF errors that occurred after PR #12884.
+func TestUnpadReaderPartialPiece(t *testing.T) {
+	actualUnpadded := abi.UnpaddedPieceSize(127 * 100) // 100 chunks = 12700 bytes
+	declaredSize := abi.PaddedPieceSize(128 * 128)     // Declare 128 chunks but only have 100
+
```
[the remark this answers is not in the document]
Comment (rvagg) [reply storage/sealer/fr32/readers_test.go:207]: no, this is wrong, and we don't need more clarity here IMO

**Review by tediou5 — APPROVED** [review]
Oh, yes. Good catch.

**Review by LexLuthr — APPROVED** [review]
